File: android/app/src/main/java/dev/terminaldeck/android/signin/ServerSignIn.kt
package dev.terminaldeck.android.signin

import android.util.Log
import dev.terminaldeck.android.crypto.HandshakeInitiator
import dev.terminaldeck.android.crypto.SealedChannel
import dev.terminaldeck.android.crypto.SealedException
import dev.terminaldeck.android.crypto.StaticKeyPair
import dev.terminaldeck.android.protocol.ClientFrames
import dev.terminaldeck.android.protocol.ClientMessage
import dev.terminaldeck.android.protocol.DeviceDescriptor
import dev.terminaldeck.android.protocol.EnrollMethod
import dev.terminaldeck.android.protocol.Protocol
import dev.terminaldeck.android.protocol.RelayWire
import dev.terminaldeck.android.protocol.ServerFrames
import dev.terminaldeck.android.transport.toHttpish
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.withTimeoutOrNull
import okhttp3.OkHttpClient
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import okio.ByteString.Companion.toByteString
import java.util.concurrent.TimeUnit

/**
 * The one-line install for a bare server, from `HEADLESS.md`.
 *
 * A phone cannot SSH, so it cannot install a host itself the way the desktop clients can — it shows
 * this command for a person to run on the machine, and then sign in. The browser client shows the
 * same line for the same reason; see `INSTALL_COMMAND` in `pwa/src/signin.ts`.
 *
 * Copied rather than derived: the domain is the marketing site's, not the app's `BRAND`, and
 * `HEADLESS.md` is the source that ships it.
 */
const val INSTALL_COMMAND = "curl -fsSL https://terminaldeck.dev/install.sh | sh"

/**
 * One sign-in, on one socket of its own.
 *
 * ## Why this is not the transport
 *
 * [dev.terminaldeck.android.transport.WebSocketDeckTransport] is a socket that is *kept*: it
 * reconnects on a backoff, it survives a phone going into a pocket, and it exists for a machine
 * this app already has a credential for. A sign-in is the opposite of all three. It happens once,
 * it must not be retried on its own — a repeated login is what a rate limiter is for, and spending
 * somebody's attempts without being asked is how a person gets locked out of their own server — and
 * at the moment it runs there is no credential for the transport to hold.
 *
 * So a sign-in opens its own short socket, finishes or fails, and hangs up. What it leaves behind is
 * an ordinary credential in the vault, and from then on the ordinary transport does what it always
 * did. Nothing in the connected path is special-cased for a server that arrived this way, which is
 * the point: a signed-in server is a paired machine, and every screen already knows what one of
 * those is.
 *
 * ## Why the durable device key, and not a throwaway
 *
 * [dev.terminaldeck.android.pairing.Rendezvous] dials with a throwaway pair, because nothing at the
 * far end of a code lookup stores who dialled. Here the opposite is true and it is load bearing:
 * `enrollDevice` on the server writes the **handshake's public key** into the device row it mints,
 * and `knowsDeviceKey` is what admits this phone on every connection afterwards. Signing in with a
 * throwaway would mint a device bound to a key this phone then throws away — a credential that
 * works exactly once, on this socket, and is refused by the same server a second later with nothing
 * on either side to explain it.
 *
 * ## The one door a server has to leave open, and what it looks like when it has not
 *
 * A server refuses the Noise handshake outright from a key it does not know, before any reply
 * exists — `isKnownDevice` in `server.ts`. A phone signing in for the first time is by definition
 * such a key, so this works only because a host that serves sign-in lets an unknown key through the
 * handshake so it can send `enroll`. A host that does not serve sign-in therefore does not refuse
 * with a sentence: it says **nothing at all** and the channel closes. That silence is the single
 * most likely failure here, and [Result.Unreachable] is worded for it rather than for a network.
 */
object ServerSignIn {

    sealed interface Result {

        data class SignedIn(
            val credential: String,
            val deviceId: String,
            val deviceName: String,
            /** The server's own capability list and sessions, from the welcome the hello earned. */
            val welcome: dev.terminaldeck.android.protocol.ServerMessage.Welcome,
        ) : Result

        /** The server answered and said no. Its sentence, not one made up here. */
        data class Refused(val sentence: String) : Result

        /** Nothing usable came back. The address, the network, or a server not serving sign-in. */
        data class Unreachable(val sentence: String) : Result
    }

    /**
     * How long a sign-in may take before it is called a failure.
     *
     * Longer than a pairing lookup, and deliberately: the server runs a real SSH probe against its
     * own sshd and then a memory-hard hash to mint the credential, and `server.ts` re-arms its own
     * no-hello timer around both because they outlast an ordinary greeting. A ceiling below the
     * server's own would report a failure for work that then succeeds, leaving a device row minted
     * on a server the phone believes it never reached.
     */
    const val TIMEOUT_MS = 45_000L

    /**
     * Everything one sign-in needs, in one value.
     *
     * A record rather than six parameters because it is also the seam: `DeckViewModel` takes a
     * `suspend (Request) -> Result` so a test can put a fixture in this object's place, and a
     * six-argument function type at that boundary is one nobody can read at the call site.
     *
     * [secret] is a password or a private-key PEM. It lives for the length of one call: it is put
     * on the wire once, inside the sealed channel, and is referenced nowhere afterwards — not
     * stored, not logged, and deliberately not kept on any view-model field.
     */
    data class Request(
        val address: ServerAddress,
        val username: String,
        val secret: String,
        val method: EnrollMethod,
        /** This phone's durable identity. See the header — a throwaway here mints a dead credential. */
        val identity: StaticKeyPair,
        val deviceName: String,
    )

    /**
     * [run] as a plain `suspend (Request) -> Result`, for a caller that wants a function reference.
     *
     * The same trick, and the same reason, as `Rendezvous.lookupAt`: a method reference to a
     * function with default arguments is not that type, and a lambda in a default parameter is a
     * closure the reader has to open the view model to understand.
     */
    suspend fun signIn(request: Request): Result = run(request)

    /**
     * Sign in to a server and come back with a credential, or with a sentence.
     *
     * `client` is a seam for the tests, exactly as it is on [dev.terminaldeck.android.pairing.Rendezvous.lookup]:
     * a unit test that used the real one would open a socket to the public relay from whatever
     * machine ran the suite.
     */
    suspend fun run(
        request: Request,
        client: OkHttpClient = defaultClient(),
        timeoutMs: Long = TIMEOUT_MS,
    ): Result {
        val address = request.address
        val username = request.username
        val secret = request.secret
        val method = request.method
        val identity = request.identity
        val deviceName = request.deviceName
        val settled = CompletableDeferred<Result>()

        // `okhttp3.Request` spelled out: [Request] in this scope is this object's own, and a socket
        // builder that silently resolved to the wrong one is the kind of collision worth naming.
        val httpRequest = try {
            okhttp3.Request.Builder()
                .url("${address.relayUrl.toHttpish().trimEnd('/')}$GUEST_PATH?host=${address.hostId}")
                .build()
        } catch (e: IllegalArgumentException) {
            return Result.Unreachable(BAD_RELAY)
        }

        val handshake = HandshakeInitiator(identity, address.hostKey)
        var channel: SealedChannel? = null
        var exchange: EnrollExchange? = null

        val socket = client.newWebSocket(httpRequest, object : WebSocketListener() {

            override fun onOpen(webSocket: WebSocket, response: Response) {
                // Message one, binary, with the version byte in front of it. That framing is the
                // difference between a handshake that is byte-perfect and one that is one byte
                // short — see [RelayWire].
                webSocket.send(RelayWire.withSealedVersion(handshake.message).toByteString())
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                val live = channel
                if (live == null) {
                    val opened = RelayWire.readSealedHandshake(
                        bytes.toByteArray(),
                        RelayWire.HANDSHAKE_REPLY_BYTES,
                    )
                    val reply = when (opened) {
                        is RelayWire.SealedOpen.Ok -> opened.message
                        is RelayWire.SealedOpen.Refused -> {
                            val sentence = if (opened.reason is RelayWire.Refusal.WrongVersion) {
                                VERSION_MISMATCH
                            } else {
                                NOT_A_HANDSHAKE
                            }
                            settled.complete(Result.Unreachable(sentence))
                            webSocket.cancel()
                            return
                        }
                    }
                    val established = try {
                        handshake.finish(reply)
                    } catch (e: SealedException) {
                        // The far end does not hold the private half of the key in the address.
                        // Either the address is not this server's, or something else answered for
                        // it — which is the exact attack carrying the key is meant to refuse.
                        settled.complete(Result.Unreachable(WRONG_KEY))
                        webSocket.cancel()
                        return
                    }
                    channel = established

                    // The exchange is built only now: it opens by sending, and there was nothing to
                    // send through until this moment.
                    val driver = EnrollExchange(
                        send = { frame -> writeSealed(webSocket, established, frame, settled) },
                        onOutcome = { outcome -> settled.complete(outcome.asResult()) },
                    )
                    exchange = driver
                    driver.start(
                        EnrollExchange.Input(
                            username = username,
                            secret = secret,
                            method = method,
                            device = DeviceDescriptor(name = deviceName, platform = PLATFORM),
                        )
                    )
                    return
                }

                val text = try {
                    live.receiveText(bytes.toByteArray())
                } catch (e: SealedException) {
                    // A frame that will not open is corruption or an impostor, and the channel's
                    // counters are the only truth about ordering — there is nothing to resynchronise
                    // to, so the sign-in ends here.
                    settled.complete(Result.Unreachable(CHANNEL_BROKEN))
                    webSocket.cancel()
                    return
                }
                when (val parsed = ServerFrames.parse(text)) {
                    is ServerFrames.Result.Bad -> Log.d(TAG, "dropped a frame during sign-in: ${parsed.reason}")
                    is ServerFrames.Result.Ok -> exchange?.receive(parsed.message)
                }
            }

            /**
             * Text frames are refused rather than ignored: everything on this socket is ciphertext,
             * so a text frame means the far end is not a relay carrying a sealed channel.
             */
            override fun onMessage(webSocket: WebSocket, text: String) {
                settled.complete(Result.Unreachable(NOT_A_RELAY))
                webSocket.cancel()
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                webSocket.close(Protocol.Close.NORMAL, null)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                // A refusal arrives as an `error` frame and *then* a close, so by the time this runs
                // the exchange has usually already settled with the server's own words. Completing
                // is a no-op on an already-settled deferred, which is what makes the ordering safe.
                settled.complete(closedResult(channel != null))
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                // The exception's message names hosts and ports and would end up on a screen; it is
                // logged instead. See the same decision in `WebSocketDeckTransport`.
                Log.d(TAG, "sign-in socket failed", t)
                settled.complete(closedResult(channel != null))
            }
        })

        return try {
            withTimeoutOrNull(timeoutMs) { settled.await() } ?: Result.Unreachable(TOO_SLOW)
        } finally {
            socket.cancel()
        }
    }

    /* -------------------------------------------------------------------- plumbing -- */

    private fun writeSealed(
        socket: WebSocket,
        channel: SealedChannel,
        frame: ClientMessage,
        settled: CompletableDeferred<Result>,
    ) {
        try {
            socket.send(channel.sendText(ClientFrames.encode(frame)).toByteString())
        } catch (e: SealedException) {
            settled.complete(Result.Unreachable(CHANNEL_BROKEN))
            socket.cancel()
        }
    }

    private fun EnrollExchange.Outcome.asResult(): Result = when (this) {
        is EnrollExchange.Outcome.SignedIn -> Result.SignedIn(
            credential = credential,
            deviceId = deviceId,
            deviceName = deviceName,
            welcome = welcome,
        )
        is EnrollExchange.Outcome.Refused -> Result.Refused(sentence)
    }

    /**
     * What a socket that closed means, and why the two answers are different.
     *
     * Before the handshake, silence is the shape a server that does not serve sign-in has: it
     * refuses an unknown device key without producing a reply, on purpose, because saying which
     * check failed would hand a hostile relay an oracle. After the handshake it is an ordinary
     * dropped connection, and saying anything about sign-in availability would be a guess.
     */
    private fun closedResult(sealed: Boolean): Result =
        Result.Unreachable(if (sealed) LOST else NO_ANSWER)

    fun defaultClient(): OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .build()

    private const val TAG = "TerminalDeck"

    /** The relay's guest endpoint, from `relay/src/rendezvous.ts`. */
    private const val GUEST_PATH = "/v1/join"

    private const val PLATFORM = "android"

    private const val BAD_RELAY = "That relay address cannot be opened. Check the wss:// part of the address."

    /**
     * Silence before the handshake, and why it names three possibilities rather than one.
     *
     * A server refuses a handshake by *saying nothing* — it will not tell a caller which check
     * failed, because that would hand a hostile relay an oracle. So all three of these arrive here
     * identically and this client genuinely cannot tell them apart: nothing is running there, it is
     * running but does not serve sign-in, or the key in the pasted address is not that server's, in
     * which case its own handshake never completes. Naming one of the three would be a guess, and
     * the wrong guess sends somebody to check the wrong thing.
     */
    const val NO_ANSWER =
        "That server did not answer this phone. It may not be running Terminal Deck right now, it " +
            "may not offer sign-in, or the address may not be that server's — check it is up, then " +
            "check you pasted the whole address."

    const val LOST = "The connection to that server dropped part-way through signing in. Try again."

    const val TOO_SLOW =
        "That server did not finish checking the sign-in in time. It may still be working — wait a " +
            "moment before trying again."

    const val WRONG_KEY =
        "That server could not prove it holds the key in that address. Check you pasted the whole " +
            "address, and that it came from the server itself."

    private const val VERSION_MISMATCH =
        "That server and this app speak different sealed-channel versions. Update whichever is older."

    private const val NOT_A_HANDSHAKE =
        "That server's first answer was not a sealed handshake. Check the relay address in what you pasted."

    private const val NOT_A_RELAY = "The relay sent something this app does not understand."

    private const val CHANNEL_BROKEN = "The connection to that server could not be verified."
}

File: ios/TerminalDeck/Transport/ServerSignIn.swift
/**
 * Signing this phone into a server, over a socket this object opens itself.
 *
 * ## What was missing, and what this is
 *
 * `SignInLink.swift` has been the tested client half of `enroll` since wave 1 —
 * the fixed frame sequence, driven, with no socket of its own. Its header says
 * so plainly: *"it does not open the socket, and it does not know how the sealed
 * channel to a first-contact host was established — that is the rendezvous
 * layer's problem, and a real one."* Nothing in the app ever solved that
 * problem, so the driver had no caller and the phone had no way to sign in to
 * anything. That is the whole of the gap 0.10.0 shipped with.
 *
 * This is the missing half: it takes a **server address** (see
 * `ServerAddress.swift` for why a host id alone cannot start a handshake), opens
 * a sealed relay channel to it with **this phone's own device key**, drives
 * `SignIn` across it, and writes the credential the server mints.
 *
 * ## The device key, not a throwaway
 *
 * `RendezvousLookup` dials with a throwaway pair, deliberately, because nothing
 * at the far end of a rendezvous stores who dialled. The opposite is true here.
 * `enroll.ts` binds the minted device row to `connection.peerPublicKey` — the
 * handshake's key — and `device-auth.ts` refuses any later handshake from a key
 * it does not know. So a sign-in run on a throwaway key would mint a device this
 * phone can never present again: a screen that says "signed in" over a machine
 * that will refuse the next connection. The durable key is the identity being
 * enrolled, so the durable key is what dials.
 *
 * ## Why it reconnects instead of keeping the socket
 *
 * The exchange ends on a `welcome`, which means this socket is *already* an
 * authenticated session. It is still torn down, and the machine is brought up
 * again through the ordinary `HostLink`. Keeping it would mean a second, special
 * path into every screen in the app — one live socket that did not come from a
 * `LiveTransport` and has none of its reconnect, heartbeat or credential
 * handling — for the sake of one round trip. Pairing has always taken the same
 * shape (`DeckModel.pairAsync` writes the record and starts the link), and this
 * matching it is what keeps the end state of the two doors identical.
 *
 * The welcome is still waited for rather than skipped. Stopping at `enrolled`
 * would be saving a credential nothing has ever spent, and calling that success.
 *
 * ## The secret is used and dropped
 *
 * The SSH password or key is held for exactly as long as the exchange, and
 * `forgetSecret` runs on every path out of it — success, refusal, timeout,
 * cancel. It is never written to the Keychain, to `UserDefaults`, or to any
 * store: what replaces it is the minted device credential, which is the point of
 * the whole ceremony. The server does the same on its side — `enroll.ts`: *"the
 * secret is used for the one probe and referenced nowhere after."*
 */

import Foundation
import Observation

@MainActor
@Observable
final class ServerSignIn {

    /**
     * What is happening, in the only five states this can honestly be in.
     *
     * The three middle ones are separate because they fail differently and a
     * person can act on the difference: nothing answering at the address is not
     * the same as a server that answered and is checking a password against its
     * own sshd, which is not the same as one that has minted a credential and is
     * being asked to honour it.
     */
    enum Phase: Equatable {
        /// Nothing in flight. The form is the screen.
        case editing
        /// Opening the socket and running the sealed handshake.
        case reaching
        /// `enroll` is out; the server is checking the login against its own sshd.
        case verifying
        /// It minted a credential; this is the `hello` that spends it.
        case joining
        /// Done. The machine is in the list and its sessions are usable.
        case signedIn(hostId: String, name: String)
        case failed(Failure)
    }

    /**
     * A refusal a person can do something about.
     *
     * Two fields rather than one sentence, because the headline is what went
     * wrong and the advice is what to do next, and collapsing them produces the
     * paragraph this product does not put on screens. Never a spinner that ends
     * nowhere: every path out of `reaching`, `verifying` and `joining` lands on
     * one of these or on `signedIn`.
     */
    struct Failure: Equatable {
        let headline: String
        let advice: String
    }

    private(set) var phase: Phase = .editing

    /// Whether something is in flight. Drives the button, and the guard that
    /// stops a second tap starting a second socket.
    var isBusy: Bool {
        switch phase {
        case .reaching, .verifying, .joining: return true
        case .editing, .signedIn, .failed: return false
        }
    }

    /**
     * The wait before the sealed channel is declared dead.
     *
     * Longer than the socket's own 20-second request timeout, so that a relay
     * that accepts and never answers is reported by the carrier's own words
     * rather than by this stopwatch — the carrier knows things this does not,
     * such as whether it got far enough to put a handshake on the wire.
     */
    private static let reachTimeout: TimeInterval = 25

    /**
     * The wait for an answer to `enroll`.
     *
     * Generous on purpose. The server's answer is a real SSH login against its
     * own sshd — `ssh-verify.ts` — behind a two-probe gate, and a phone that
     * gave up at ten seconds would report a working server as broken every time
     * somebody else was signing in at the same moment.
     */
    private static let verifyTimeout: TimeInterval = 45

    private let credentials: CredentialStore
    private let device: DeviceDescriptor
    private let makeCarrier: (DeckEndpoint, StaticKeyPair) -> Carrier
    private let onSignedIn: (StoredCredential) -> Void

    private var carrier: Carrier?
    private var driver: SignIn?
    private var endpoint: DeckEndpoint?
    /// Held for the length of the exchange and nothing longer. See the header.
    private var secret = ""
    /// The code on the last `error` frame, kept because `SignIn` collapses an
    /// error to its sentence and the code is what says whether this server
    /// *refused* the login or cannot offer sign-in at all.
    private var refusalCode: ProtocolErrorCode?
    /// Bumped on every attempt, so a timer armed by an abandoned one does nothing.
    private var generation = 0

    /// `makeCarrier` is the one seam for the network: the tests drive this
    /// against a scripted carrier rather than a relay.
    init(credentials: CredentialStore,
         device: DeviceDescriptor,
         makeCarrier: @escaping (DeckEndpoint, StaticKeyPair) -> Carrier = LiveTransport.defaultCarrier,
         onSignedIn: @escaping (StoredCredential) -> Void) {
        self.credentials = credentials
        self.device = device
        self.makeCarrier = makeCarrier
        self.onSignedIn = onSignedIn
    }

    // MARK: - Driving one sign-in

    /**
     * Sign in, or say why not.
     *
     * The three refusals before a socket is opened are all about the form, and
     * each is answered on the form rather than by dialling something that cannot
     * work: an address that does not parse, a missing username, a missing
     * secret. Only after those does anything touch the network.
     */
    func submit(address: String, username: String, secret: String, method: EnrollMethod) {
        guard !isBusy else { return }

        let endpoint: DeckEndpoint
        switch ServerAddress.parse(address) {
        case let .failure(error):
            phase = .failed(Failure(headline: "That server address was not readable.",
                                    advice: error.detail))
            return
        case let .success(parsed):
            endpoint = parsed
        }

        let account = username.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !account.isEmpty else {
            phase = .failed(Failure(
                headline: "That sign-in needs a username.",
                advice: "The account you would use to SSH into that server."))
            return
        }
        // Not trimmed. A password may legitimately begin or end with a space,
        // and a private key ends with a newline that is part of the format.
        guard !secret.isEmpty else {
            phase = .failed(Failure(
                headline: method == .password ? "That sign-in needs a password." : "That sign-in needs a key.",
                advice: method == .password
                    ? "The password for that account on that server."
                    : "Paste the private key for that account, including its BEGIN and END lines."))
            return
        }

        generation &+= 1
        let epoch = generation
        self.endpoint = endpoint
        self.secret = secret
        refusalCode = nil
        phase = .reaching

        let carrier = makeCarrier(endpoint, credentials.deviceKeys())
        self.carrier = carrier
        carrier.onEvent = { [weak self] event in
            guard let self, self.generation == epoch else { return }
            switch event {
            case .ready:
                self.opened(account: account, method: method)
            case let .text(text):
                self.received(text)
            case let .closed(close):
                self.closed(close)
            }
        }
        carrier.open()

        after(Self.reachTimeout, epoch: epoch) { [weak self] in
            guard let self, case .reaching = self.phase else { return }
            self.fail(Failure(
                headline: "That server did not answer.",
                advice: "The address was reached and nothing on the other end replied. Check the "
                    + "server is awake and running \(Brand.name), then try again."))
        }
    }

    /// Give up on an exchange in flight. The screen's Cancel, and the only thing
    /// that stops a sign-in early.
    func cancel() {
        guard isBusy else { return }
        forgetSecret()
        generation &+= 1
        teardown()
        phase = .editing
    }

    /// Put the form back after a refusal, so the person can fix one field rather
    /// than being left on a dead end.
    func edit() {
        guard !isBusy else { return }
        phase = .editing
    }

    // MARK: - The exchange

    private func opened(account: String, method: EnrollMethod) {
        guard case .reaching = phase else { return }
        phase = .verifying

        let driver = SignIn(send: { [weak self] message in
            _ = self?.carrier?.send(WireCodec.encode(message))
        }, onOutcome: { [weak self] outcome in
            self?.settle(outcome)
        })
        self.driver = driver
        driver.start(SignInInput(username: account, secret: secret, method: method, device: device))

        let epoch = generation
        after(Self.verifyTimeout, epoch: epoch) { [weak self] in
            guard let self else { return }
            switch self.phase {
            case .verifying, .joining:
                self.fail(Failure(
                    headline: "That server has not finished checking the sign-in.",
                    advice: "It checks the login against its own SSH, and that has not come back. "
                        + "Try again in a moment."))
            default:
                return
            }
        }
    }

    private func received(_ raw: String) {
        guard case let .ok(message, _) = WireCodec.decode(raw) else {
            // A frame this build cannot read is dropped rather than fatal: the
            // driver ignores everything before the welcome anyway, and a message
            // type added on the server side must not end a sign-in.
            return
        }
        // Read before the driver sees it, because the driver keeps the sentence
        // and throws the code away — and the code is the difference between a
        // wrong password and a server that has no sign-in to offer.
        if case let .error(code, _) = message { refusalCode = code }
        if case .enrolled = message, case .verifying = phase { phase = .joining }
        driver?.receive(message)
    }

    private func settle(_ outcome: SignInOutcome) {
        switch outcome {
        case let .ok(token, deviceId, deviceName):
            guard let endpoint else { return }
            // A machine signed into twice keeps the name its owner gave it, the
            // same rule a re-pair follows.
            let existing = credentials.load(endpoint.hostId)
            let stored = StoredCredential(endpoint: endpoint,
                                          token: token,
                                          kind: .device,
                                          deviceId: deviceId,
                                          deviceName: deviceName,
                                          pairedAt: existing?.pairedAt ?? Date(),
                                          nickname: existing?.nickname,
                                          hostName: existing?.hostName)
            // Written before anything is said on screen. A screen that reported
            // success ahead of the write would be a machine that vanishes when
            // the app is killed a second later — with a device row on the server
            // that nothing can ever present again.
            credentials.save(stored)
            forgetSecret()
            generation &+= 1
            teardown()
            phase = .signedIn(hostId: stored.hostId, name: stored.label)
            onSignedIn(stored)

        case let .failed(message):
            fail(refusal(message))
        }
    }

    /**
     * What the server's refusal means, in words that name the next move.
     *
     * The sentence is the server's own wherever it has one — those are written
     * for a person and `enroll.ts` deliberately collapses a bad password and a
     * rate-limited address into one of them — and the code decides which
     * headline it sits under.
     */
    private func refusal(_ message: String) -> Failure {
        let said = message.isEmpty ? nil : message
        switch refusalCode {
        case .some(.unavailable):
            /*
             * The headline used to read "That server does not offer sign-in",
             * and on 2026-08-22 that sentence was a lie for an entire evening.
             *
             * `unavailable` is not one thing. The host sends it when sign-in is
             * genuinely switched off, when its own sshd did not answer the
             * loopback probe, when it is out of device slots, and when it could
             * not write the new device row. Three of those four are a running
             * server that serves sign-in perfectly — and the one that happened
             * was an sshd on a non-standard port, with the phone insisting the
             * feature was not built into the machine.
             *
             * The wire cannot tell them apart (the code is the same and a fifth
             * error code would print as "unknown" on every shipped client), so
             * the headline says only what is true of all four and the host's own
             * sentence carries the rest. Since 2026-08-23 that sentence names
             * the port it dialled and the variable that moves it.
             */
            return Failure(
                headline: "That server could not sign this device in.",
                advice: said ?? "Pair it with a code from its own screen instead.")
        case .some(.version):
            return Failure(
                headline: "The two ends do not speak the same protocol.",
                advice: said ?? "Update whichever of the app and the server is older.")
        case .some(.badMessage), .some(.tooLarge), .some(.unknownSession), .none:
            // An older server does not know the word `enroll` at all: it refuses
            // the frame as a bad message rather than as a bad login. Saying "that
            // sign-in was refused" there would send somebody to check a password
            // that was never read.
            return Failure(
                headline: "That server is too old to sign a phone in.",
                advice: "It did not recognise the sign-in. Update \(Brand.name) on the server, or "
                    + "pair it with a code from its own screen instead.")
        case .some(.unauthorized), .some(.unauthenticated):
            return Failure(
                headline: "That sign-in was refused.",
                advice: said ?? "Check the username, and the password or key, then try again.")
        }
    }

    private func closed(_ close: CarrierClose) {
        switch phase {
        case .editing, .signedIn, .failed:
            // Either this side hung up on purpose, or the exchange has already
            // settled and this is the socket going away behind it.
            return
        case .reaching, .verifying, .joining:
            break
        }

        if close.beforeReady {
            // The carrier's own sentence, because it knows things this does not
            // — whether it reached the relay at all, and whether the handshake
            // was ever put on the wire. A server older than sign-in refuses the
            // handshake outright (`isKnownDevice` in `server.ts` lets a stranger
            // through only when sign-in is served), so it arrives here rather
            // than as a refusal, and the advice has to name that too.
            fail(Failure(
                headline: "Could not reach that server.",
                advice: (close.detail.map { $0 + " " } ?? "")
                    + "Check the server is awake and running \(Brand.name). A server too old to "
                    + "offer sign-in refuses a first connection the same way — pair that one with a "
                    + "code instead."))
            return
        }

        fail(Failure(
            headline: "That server hung up part way through.",
            advice: (close.detail.map { $0 + " " } ?? "")
                + "Nothing was signed in. Try again."))
    }

    private func fail(_ failure: Failure) {
        forgetSecret()
        generation &+= 1
        teardown()
        phase = .failed(failure)
    }

    /**
     * Drop the SSH secret.
     *
     * Called on every path out of an exchange. Overwritten rather than left to
     * be reassigned later: a `String` this class stops referencing is a `String`
     * whose bytes are wherever the allocator left them, and the honest thing
     * available in Swift is to stop holding it at the first moment it is no
     * longer needed. Nothing else in this app is ever handed it.
     */
    private func forgetSecret() {
        secret = ""
    }

    private func teardown() {
        driver = nil
        carrier?.onEvent = nil
        carrier?.close()
        carrier = nil
    }

    private func after(_ delay: TimeInterval, epoch: Int, _ body: @escaping @MainActor () -> Void) {
        Task { @MainActor [weak self] in
            try? await Task.sleep(for: .seconds(delay))
            guard let self, self.generation == epoch else { return }
            body()
        }
    }
}
